### ISSUE TITLE
lldb-vscode not configured properly on OS:Windows?

### ISSUE THREAD

**Issue by DrKGD** [mfussenegger/nvim-dap#260]
### Debug adapter definition and debug configuration

I have been trying using this plugin with clang/gcc compilers under mingw toolchain on **OS:Windows 10**.
Note that I tried both [LLVM](https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.1`) 64-bit version and [MinGW](https://packages.msys2.org/package/mingw-w64-clang-x86_64-lldb?repo=clang64) `lldb-vscode` executables with the same exact result.

I am not an expert in the matter but I am able to debug the _relatively simple code_ using command line `lldb`. As for the configuration details, I have just followed the READMEs ones.

```
dap.adapters.lldb = {
	type	= 'executable',
	command = 'lldb-vscode',
	name	= 'lldb',
}

dap.configurations.cpp = {
	{
		name = "Launch",
		type = "lldb",
		request = "launch",
		cwd = '${workspaceFolder}',
		stopOnEntry = false,
		args = {},
		runInTerminal = false,
		program = function() return vim.fn.input('Path to executable:', vim.fn.getcwd() .. '/', 'file') end,
	}
}

dap.set_log_level('TRACE')
dap.configurations.c	= dap.configurations.cpp
dap.configurations.rust = dap.configurations.cpp
```

I also have installed dapui, disabling it does not affect the end result.

Finally, I am compiling this simplified example with the following flags (as I read that dwarf-5 could probably freak out lldb-vscode in another issue #186 even though it seems unreleated)

`clang++ -gdwarf-4 -target x86_64-w64-windows-gnu main.cpp -o main.exe`

```cpp
#include <iostream>

int main(){
	int a = 23;
	int b = 34;
	int c = a + b;

	std::cout<< "Sample program\n";
	if (true) std::cout<< "Ok.";

	return 0;
}
```

It is **HIGHLY** due to a misconfiguration on my end of some sort. 

### Debug adapter version

lldb-vscode (lldb 12.0.1)

### Steps to Reproduce

1. Compile any cpp source with debug symbols (in this case, I am forcing -gdwarf-4 because I am unsure what clang/gcc uses by default `clang++ -gdwarf-4 -target x86_64-w64-windows-gnu main.cpp -o main.exe`)
2. In neovim, add a breakpoint using `dap.toggle_breakpoint()`
3. In neovim, run `dap.continue()` and fill in the executable name accordingly (main.exe)

### Expected Result

Active debug session, just like in a lldb instance 

```
C:\Users\drkgd\Projects\NeovimConfiguration>clang++ main.cpp -o main.exe -gdwarf-4 -target x86_64-w64-windows-gnu

C:\Users\drkgd\Projects\NeovimConfiguration>lldb main.exe
(lldb) target create "main.exe"
Current executable set to 'C:\Users\drkgd\Projects\NeovimConfiguration\main.exe' (x86_64).
(lldb) b main
Breakpoint 1: where = main.exe`main + 22 at main.cpp:4:6, address = 0x0000000140001496
(lldb) r
Process 18236 launched: 'C:\Users\drkgd\Projects\NeovimConfiguration\main.exe' (x86_64)
Process 18236 stopped
* thread #1, stop reason = breakpoint 1.1
    frame #0: 0x00007ff655fa1496 main.exe`main at main.cpp:4:6
   1    #include <iostream>
   2
   3    int main(){
-> 4            int a = 23;
   5            int b = 34;
   6            int c = a + b;
   7
(lldb) c
Process 18236 resuming
Process 18236 exited with status = 0 (0x00000000)
(lldb)
```

### Actual Result

Dap opens a new vim buffer with (presumably) an assembly dump of the entire executable, nothing too obvious but a warn in the log(s) 

```
[ WARN ] 2021-08-02T14:59:09Z+0200 ] ...data\site\pack\packer\start\nvim-dap\lua\dap\session.lua:518 ]	"No callback for "	{
  body = {
    allThreadsContinued = true
  },
  command = "continue",
  request_seq = 14,
  seq = 0,
  success = true,
  type = "response"
}
```

https://user-images.githubusercontent.com/44092476/127865718-24f700a7-51ac-463d-a0bd-ac5bfee5d7cf.mp4

[dap.log](https://github.com/mfussenegger/nvim-dap/files/6917046/dap.log)

Also I can ensure that nvim-dap, along with dapui, works perfectly fine on other adapters, (e.g. nlua with osv).
Thanks in advance!

**Comment by mfussenegger** [comment]
The log indicates you tried to debug a `test2.go` file. Are you sure you uploaded the right log?

**Comment by DrKGD** [comment]
Uh, quite sure that I have uploaded the right log file, not really into .go coding so that's out of question?

Anyway, just for the sake of it, I recreated the log by following the same procedure, and renamed the file. I also pasted the content onto pastebin, just in case. [dap-08082021.log](https://github.com/mfussenegger/nvim-dap/files/6950950/dap-08082021.log)

[Pastebin](https://pastebin.com/W7uVzrpV), pw: **dapdebug**

**Comment by mfussenegger** [comment]
Looks like an exception occurs:

```
[ DEBUG ] 2021-08-08T16:38:01Z+0200 ] ...data\site\pack\packer\start\nvim-dap\lua\dap\session.lua:510 ]	{
  body = {
    breakMode = "always",
    description = "Exception 0x80000003 encountered at address 0x7ff961f00770",
    exceptionId = "exception"
  },
  command = "exceptionInfo",
  request_seq = 5,
  seq = 0,
  success = true,
  type = "response"
}
```

And it then immediately jumps to the disassembly where it occurred. 
Could be a lldb-vscode issue / Windows issue?

I've the impression that https://github.com/mfussenegger/nvim-dap/wiki/C-C---Rust-(gdb-via--vscode-cpptools) currently is  more reliable.

**Comment by DrKGD** [comment]
Thanks for the heads-up, even though I cannot really follow that procedure either, I assume that `cpptools-win32.vsix` is the right version [Windows 64-bit & 32-bit](https://github.com/microsoft/vscode-cpptools#offline-installation) and I can't really follow your guide as it `cpptools-win32` is missing `extension/debugAdapters/mono.<system-<platform>` and `extension/debugAdapters/OpenDebugAD7` is a bash script.

My best guess is that I should just run `extension\debugAdapters\bin\OpenDebugAD7.exe`, which is an executable, but as I though I am missing something: running it either in a command prompt or via dap it hangs indefinitely on `waiting for v8 protocol on stdin/stdout`.

```
[ DEBUG ] 2021-08-09T21:35:29Z+0200 ] ...data\site\pack\packer\start\nvim-dap\lua\dap\session.lua:742 ]	"request"	{
  arguments = {
    adapterID = "nvim-dap",
    clientId = "neovim",
    clientname = "neovim",
    columnsStartAt1 = true,
    linesStartAt1 = true,
    locale = "en_US",
    pathFormat = "path",
    supportsRunInTerminalRequest = true,
    supportsVariableType = true
  },
  command = "initialize",
  seq = 0,
  type = "request"
}
[ ERROR ] 2021-08-09T21:35:29Z+0200 ] ...data\site\pack\packer\start\nvim-dap\lua\dap\session.lua:650 ]	"stderr"	{
  command = "C:\\Users\\drkgd\\.config\\nvim\\exe\\cpptools-win32\\extension\\debugAdapters\\bin\\OpenDebugAD7.exe",
  type = "executable"
}	"waiting for v8 protocol on stdin/stdout\r\n"

```

Reading through `extension/debugAdapters/OpenDebugAD7`, I suspect that `OpenDebugAD7.exe` requires parameters of _some_ kind.

Do you have any suggestion on the matter? 
Thanks again for your availability!

**Comment by mfussenegger** [comment]
I think under windows you always need to set the `MIMode` and `MIDebuggerPath` properties. For example:


```lua
dap.configurations.cpp = {
  {
    name = "Launch file",
    type = "cppdbg",
    request = "launch",
    program = function()
      return vim.fn.input('Path to executable: ', vim.fn.getcwd() .. '/', 'file')
    end,
    cwd = '${workspaceFolder}',
    stopOnEntry = true,
    MIMode = 'gdb',
    MIDebuggerPath = 'gdb.exe',
  },
}
```


`OpenDebugAD7.exe` is the right executable.

**Comment by DrKGD** [comment]
Seems I am having no luck either, the same exact _error_ message is happening `waiting for v8 protocol on stdin/stdout\r\n`

```

[ DEBUG ] 2021-08-09T23:31:01Z+0200 ] ...data\site\pack\packer\start\nvim-dap\lua\dap\session.lua:742 ]	"request"	{
  arguments = {
    adapterID = "nvim-dap",
    clientId = "neovim",
    clientname = "neovim",
    columnsStartAt1 = true,
    linesStartAt1 = true,
    locale = "en_US",
    pathFormat = "path",
    supportsRunInTerminalRequest = true,
    supportsVariableType = true
  },
  command = "initialize",
  seq = 0,
  type = "request"
}
[ ERROR ] 2021-08-09T23:31:01Z+0200 ] ...data\site\pack\packer\start\nvim-dap\lua\dap\session.lua:650 ]	"stderr"	{
  command = "C:\\Users\\drkgd\\.config\\nvim\\exe\\cpptools-win32\\extension\\debugAdapters\\bin\\OpenDebugAD7.exe",
  type = "executable"
}	"waiting for v8 protocol on stdin/stdout\r\n"

```

Following your advices, my setup is exactly the following

```
dap.adapters.cppdbg = {
	type = 'executable',
	command = 'C:\\Users\\drkgd\\.config\\nvim\\exe\\cpptools-win32\\extension\\debugAdapters\\bin\\OpenDebugAD7.exe'
}

dap.configurations.cpp = {
	{
		name = "Launch file",
		type = "cppdbg",
		request = "launch",
		program = function()
			return vim.fn.input('Path to executable: ', vim.fn.getcwd() .. '/', 'file')
		end,
		cwd = '${workspaceFolder}',
		stopOnEntry = true,
		externalConsole = false,
		MIMode = 'gdb',
		MIDebuggerPath = 'gdb.exe',
	}
}
```

**Comment by mfussenegger** [comment]
Looks like there are others who have this issue as well: https://github.com/microsoft/vscode-cpptools/issues/2382  (using vscode instead of nvim-dap).

**Comment by DrKGD** [comment]
[Arch](https://archlinux.org/packages/extra/x86_64/lldb/) lldb-vscode is working flawlessly on WSL (Windows Subsystem for Linux), as expected.
![immagine](https://user-images.githubusercontent.com/44092476/129098724-458e1454-188d-4128-9dd6-b2cd14d9d2cc.png)

_(running neovide gui client with --wsl flag)_
Seems that for now I must give up in trying to make it work on _vanilla Windows_ 🙌. Thanks anyway!

**Comment by mfussenegger** [comment]
Closing this here given that it seems to be a debug adapter issue, and not a client issue.

**Comment by mfussenegger** [comment]
vscode-cpptools should work now under windows (without WSL) with a configuration like this:

```
local dap = require('dap')
dap.adapters.cppdbg = {
  id = 'cppdbg',
  type = 'executable',
  command = 'C:\\absolute\\path\\to\\cpptools\\extension\\debugAdapters\\bin\\OpenDebugAD7.exe',
  options = {
    detached = false
  }
}
```

**Comment by ritchielrez** [comment]
> vscode-cpptools should work now under windows (without WSL) with a configuration like this:
> 
> ```lua
> local dap = require('dap')
> dap.adapters.cppdbg = {
>   id = 'cppdbg',
>   type = 'executable',
>   command = 'C:\\absolute\\path\\to\\cpptools\\extension\\debugAdapters\\bin\\OpenDebugAD7.exe',
>   options = {
>     detached = false
>   }
> }
> ```

It doesn't though for me.

**Comment by derekthecool** [comment]
I had similar issues as described. I'm running windows 11 and was able to get past the issue.

I'm using mason to install the cppdbg debugger and then I applied the options table in my config as well as adding the `MIMode` and `MIDebuggerPath` to my config.

I just used `scoop install gdb` to get gdb installed quickly. No full paths were needed. Scoop adds them to the path.

My full config here

```lua
local dap = require('dap')
dap.adapters.cppdbg = {
    id = 'cppdbg',
    type = 'executable',
    command = OS.join_path(Mason.bin, 'OpenDebugAD7' .. OS.executable_extension_alt),
    options = {
        detached = false,
    },
    MIMode = 'gdb',
    MIDebuggerPath = 'gdb.exe',
}
-- Set the C++ settings to work for C as well
dap.configurations.c = dap.configurations.cpp
```

The OS.join_path function is in my personal neovim config. It just joins paths with the correct `/` or `\`.
The other global I have is `OS.executable_extension_alt` which evaluates to `.cmd` on windows and empty on Linux (mason stuff)
